include/onnxruntime/core/graph/graph.h

```diff
@@ -343,22 +343,50 @@ class Node {
   size_t GetOutputEdgesCount() const noexcept { return relationships_.output_edges.size(); }
 
   /** Add an attribute to this Node with specified attribute name and value. */
-  void AddAttribute(const std::string& attr_name, const ONNX_NAMESPACE::AttributeProto& value);
+  void AddAttribute(std::string attr_name, const ONNX_NAMESPACE::AttributeProto& value);
+  void AddAttribute(std::string attr_name, ONNX_NAMESPACE::AttributeProto&& value);
 
-#define ADD_ATTR_INTERFACES(TypeName)                                     \
-  void AddAttribute(const std::string& attr_name, const TypeName& value); \
-  void AddAttribute(const std::string& attr_name,                         \
-                    const std::vector<TypeName>& values);
+#define ADD_ATTR_INTERFACES(TypeName)                              \
+  void AddAttribute(std::string attr_name, const TypeName& value); \
+  void AddAttribute(std::string attr_name,                         \
+                    gsl::span<TypeName const> values);
+
+#define ADD_ATTR_MOVE_INTERFACE(TypeName) \
+  void AddAttribute(std::string attr_name, TypeName&& value);
+
+  void AddAttribute(std::string attr_name, std::string value);
+  void AddAttribute(std::string attr_name, gsl::span<std::string const> values);
 
   ADD_ATTR_INTERFACES(int64_t)
   ADD_ATTR_INTERFACES(float)
-  ADD_ATTR_INTERFACES(std::string)
   ADD_ATTR_INTERFACES(ONNX_NAMESPACE::TensorProto)
-  ADD_ATTR_INTERFACES(ONNX_NAMESPACE::GraphProto)
+  ADD_ATTR_MOVE_INTERFACE(ONNX_NAMESPACE::TensorProto)
 #if !defined(DISABLE_SPARSE_TENSORS)
   ADD_ATTR_INTERFACES(ONNX_NAMESPACE::SparseTensorProto)
+  ADD_ATTR_MOVE_INTERFACE(ONNX_NAMESPACE::SparseTensorProto)
 #endif
   ADD_ATTR_INTERFACES(ONNX_NAMESPACE::TypeProto)
+  ADD_ATTR_MOVE_INTERFACE(ONNX_NAMESPACE::TypeProto)
+
+  void AddAttribute(std::string attr_name, const ONNX_NAMESPACE::GraphProto& value);
+  void AddAttribute(std::string attr_name, ONNX_NAMESPACE::GraphProto&& value);
+
+  // The below overloads are made so the compiler does not attempt to resolve
+  // C-strings with a gsl::span overloads
+  template <size_t N>
+  void AddAttribute(std::string attr_name, const char (&value)[N]) {
```

Comment:
>AddAttribute [](http://example.com/codeflow?start=7&length=12)

just curious, would we save binary size if we have const char* overloads instead?

Reply:
IDK, I though it was pretty cool to re-use the size without calling `strlen()` and the function call is going to be inlined anyway.

Reply:
Agreed to handle this if this shows up as a concern.